test/functional/legacy/060_exists_and_has_functions_spec.lua

```diff
@@ -0,0 +1,874 @@
+            echo "OK"
+          else
+            echo "FAILED"
+          endif
+          " Existing buffer dictionary.
+          let b:buffer_dict = {"xcord":100, "ycord":2}
+          echo 'b:buffer_dict: 1'
+          if exists('b:buffer_dict')
+            echo "OK"
+          else
+            echo "FAILED"
+          endif
+          " Non-existing buffer dictionary.
+          unlet b:buffer_dict
+          echo 'b:buffer_dict: 0'
+          if !exists('b:buffer_dict')
+            echo "OK"
+          else
+            echo "FAILED"
+          endif
+          " Existing buffer curly-brace variable.
+          let str = "buffer"
+          let b:curly_{str}_var = 1
+          echo 'b:curly_' . str . '_var: 1'
+          if exists('b:curly_{str}_var')
+            echo "OK"
+          else
+            echo "FAILED"
+          endif
+          " Non-existing buffer curly-brace variable.
+          unlet b:curly_{str}_var
+          echo 'b:curly_' . str . '_var: 0'
+          if !exists('b:curly_{str}_var')
+            echo "OK"
+          else
+            echo "FAILED"
+          endif
+          " Script-local tests.
+          source test60.vim
+          " Existing Vim internal variable.
+          echo 'v:version: 1'
+          if exists('v:version')
+            echo "OK"
+          else
+            echo "FAILED"
+          endif
+          " Non-existing Vim internal variable.
+          echo 'v:non_exists_var: 0'
+          if !exists('v:non_exists_var')
+            echo "OK"
+          else
+            echo "FAILED"
+          endif
+          " Function arguments.
+          function TestFuncArg(func_arg, ...)
+              echo 'a:func_arg: 1'
+              if exists('a:func_arg')
+                  echo "OK"
+              else
+                  echo "FAILED"
+              endif
+              echo 'a:non_exists_arg: 0'
+              if !exists('a:non_exists_arg')
+                  echo "OK"
+              else
+                  echo "FAILED"
+              endif
+              echo 'a:1: 1'
+              if exists('a:1')
+                  echo "OK"
+              else
+                  echo "FAILED"
+              endif
+              echo 'a:2: 0'
+              if !exists('a:2')
+                  echo "OK"
+              else
+                  echo "FAILED"
+              endif
+          endfunction
+          call TestFuncArg("arg1", "arg2")
+          echo ' g:footest#x =' g:footest#x
+          echo '   footest#F()' footest#F()
+          echo 'UndefFun()' UndefFun()
+          redir END
+      endfunction
+
+      call TestExists()
+
+      function TestHas()
+        redir >> test.out
+        for pl in ['6.9.999', '7.1.999', '7.4.123', '9.1.0', '9.9.1']
+          echo 'has patch ' . pl . ': ' . has('patch-' . pl)
+        endfor
+        redir END
+      endfunc
+      call TestHas()
+
+      delfunc TestExists
+      delfunc RunTest
+      delfunc TestFuncArg
```

Comment:
I removed this `delfunc` post-cleanup, they aren't needed since we cleanup the environment on each test case.